### ISSUE TITLE
Medusa backups in multiple S3 buckets

### ISSUE THREAD

**Issue by sandeepmallik** [thelastpickle/cassandra-medusa#409]
Hi,

I have a 40 node cluster which is under constant load. Backups to S3 is failing due to this + network is overwhelmed due to amount of data to be uploaded.

_Can I upload backups to multiple regions?_ OR _Can I upload backups to multiple buckets in same region?_ Will medusa supports this feature? I know medusa stores metadata of backups but not sure if it can support for multiple locations.

Ex:- This is what we want to achieve.

As there are 40 nodes, we will create 4 S3 buckets in multiple regions or 4 S3 buckets in same region. Will assign a bucket for 10 nodes each and use medusa to upload data. Will this work?

┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-1397) by [Unito](https://www.unito.io)
┆friendlyId: K8SSAND-1397
┆priority: Medium

**Comment by adejanovski** [comment]
Hi @sandeepmallik,

we do not have this feature on the roadmap for now.
My recommendation currently would be to use the `medusa backup-cluster` command to orchestrate the backups and control the level of parallelism of the uploads.
Data will be snapshotted on all nodes at (almost) the same time, and then uploads will be done in parallel for the number of nodes you want by using the `--parallel-uploads=` argument: https://github.com/thelastpickle/cassandra-medusa/blob/master/medusa/medusacli.py#L145-L148

You can also throttle your upload speed to S3 and number of concurrent uploads per node by setting the appropriate values in `medusa.ini`: https://github.com/thelastpickle/cassandra-medusa/blob/master/medusa-example.ini#L67-L71

**Comment by m934030039** [comment]
Hi guys,

    We are also suffering from this kind of issue that is 40+ nodes with huge backup data to uploads to S3.

    We did modify max upload speed and concurrent upload threads and it was ok but now it cannot work well after growing the number of nodes.

    We are investigating below 
          *). Traffice between EC2 instances and S3. (https://aws.amazon.com/premiumsupport/knowledge-center/s3-maximum-transfer-speed-ec2/)
          *). system average load during the Medusa backup.

@adejanovski 
   Do you have any suggestion from Medusa's prespective?

-Sheng

**Comment by sandeepmallik** [comment]
Thanks. 

Also, from uptime

**Comment by m934030039** [comment]
Hi Adejanovski,

    I am confusing aboht the C* in below image.

    Could you clarify that if we do cluster wide backup, cstar utility is NOT required or required ?
 
![image](https://user-images.githubusercontent.com/31229402/135206850-a34d6881-b8e1-4145-9325-5e03a48f3894.png)

-Sheng

**Comment by adejanovski** [comment]
> Hi guys,
> 
> ```
> We are also suffering from this kind of issue that is 40+ nodes with huge backup data to uploads to S3.
> 
> We did modify max upload speed and concurrent upload threads and it was ok but now it cannot work well after growing the number of nodes.
> 
> We are investigating below 
>       *). Traffice between EC2 instances and S3. (https://aws.amazon.com/premiumsupport/knowledge-center/s3-maximum-transfer-speed-ec2/)
>       *). system average load during the Medusa backup.
> ```
> 
> @adejanovski Do you have any suggestion from Medusa's prespective?
> 
> -Sheng

My recommendation would be the same: use `medusa backup-cluster` and limit the number of parallel uploads to match the available bandwidth.

> I am confusing about the C* in below image.
> 
> Could you clarify that if we do cluster wide backup, cstar utility is NOT required or required ?

C* stands for Cassandra, not cstar ;) 
You don't need cstar if you're using `medusa backup-cluster` or if you're using cron jobs. Since we introduced the `backup-cluster` command, I'd say there's no point in using cstar to create backups anymore.

**Comment by m934030039** [comment]
Hi Adejanovski,

     Thx for the confirmation.
     If that resolves, I will update here.
     I will make a summary for this issue and share it with you guys.

-Sheng
-Sheng

**Comment by sandeepmallik** [comment]
Hi Adejanovski,

We are seeing below error. But, the object is already present in S3. 

Not sure why medusa is complaining the same. 

![image](https://user-images.githubusercontent.com/12246569/136138150-cbecf965-6e80-4cca-b541-8b2e2a430178.png)

[2021-09-24 00:59:10,083] DEBUG: nodetool output: Requested clearing snapshot(s) for [all keyspaces] with snapshot name [medusa-s3.full_20210923]
[2021-09-24 00:59:10,084] ERROR: This error happened during the backup: Failed to find uploaded object test-tmp-cluster/cass-913.test.com/s3.full_20210923/data/test_tmp/log5-80fd21a0ebf111ea894ff37a6804a092/md-599-big-Data.db in S3

manifest.json is not present when backup failed on node. Will this file gets created only when medusa backup is successful? I was under impression that it gets created when backup starts. I feel medusa is unable to find object reference in manifest.json (which is not there) and it is failing.

**Comment by adejanovski** [comment]
@sandeepmallik, did you manage to fix this issue? It's indeed a path problem as you'll need nodetool to be in the PATH of the user running medusa.

**Comment by jsanda** [comment]
Please [add your planning poker estimate](https://app.zenhub.com/workspaces/K8ssandra-Grooming-623b197f1fcd4800110540ef/issues/thelastpickle/cassandra-medusa/409?planning-poker) with ZenHub @adejanovski

**Comment by adejanovski** [comment]
We're probably not going to support this feature anytime soon. It would be way too complex to spread backups in different buckets.
Note that backups are made per DC and distinct DCs can use different buckets, possibly in different regions.

My recommendation would be either:

- to perform more frequent differential backups so that the uploaded volume gets smaller
- to use the `medusa backup-cluster` command and tune the upload concurrency by using the `--parallel-uploads` argument

**Comment by adejanovski** [comment]
I'll close this issue as "won't do". There are ways to work around that issue as explained above (different buckets per DC, more frequent differential backups, tuning parallelism).
